records.go

```diff
@@ -34,44 +38,75 @@ func (dht *IpfsDHT) GetPublicKey(ctx context.Context, p peer.ID) (ci.PubKey, err
 		return pk, nil
 	}
 
-	// ok, try the node itself. if they're overwhelmed or slow we can move on.
-	ctxT, cancelFunc := ctxfrac.WithDeadlineFraction(ctx, 0.3)
-	defer cancelFunc()
-	if pk, err := dht.getPublicKeyFromNode(ctx, p); err == nil {
-		err := dht.peerstore.AddPubKey(p, pk)
-		if err != nil {
-			return pk, err
+	// Try getting the public key both directly from the node it identifies
+	// and from the DHT, in parallel
+	resp := make(chan pubkrs, 2)
+	go func() {
+		pubk, err := dht.getPublicKeyFromNode(ctx, p)
+		resp <- pubkrs{pubk, err}
+	}()
+
+	go func() {
+		pubk, err := dht.getPublicKeyFromDHT(ctx, p)
+		resp <- pubkrs{pubk, err}
+	}()
+
+	// Wait for one of the two go routines to return
+	// a public key (or for both to error out)
+	var err error
+	for i := 0; i < 2; i++ {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case r := <-resp:
+			if r.err == nil {
+				// Found the public key
+				err := dht.peerstore.AddPubKey(p, r.pubk)
+				if err != nil {
+					log.Warningf("Failed to add public key to peerstore for %v", p)
+				}
+				return r.pubk, nil
+			}
+			err = r.err
 		}
-		return pk, nil
 	}
 
-	// last ditch effort: let's try the dht.
-	log.Debugf("pk for %s not in peerstore, and peer failed. Trying DHT.", p)
-	pkkey := routing.KeyForPublicKey(p)
+	// Both go routines failed to find a public key
+	return nil, err
+}
 
-	val, err := dht.GetValue(ctxT, pkkey)
+func (dht *IpfsDHT) getPublicKeyFromDHT(ctx context.Context, p peer.ID) (ci.PubKey, error) {
+	// Only retrieve one value, because the public key is immutable
+	// so there's no need to retrieve multiple versions
+	pkkey := routing.KeyForPublicKey(p)
+	vals, err := dht.GetValues(ctx, pkkey, 1)
 	if err != nil {
-		log.Warning("Failed to find requested public key.")
 		return nil, err
 	}
 
-	pk, err = ci.UnmarshalPublicKey(val)
+	if len(vals) == 0 || vals[0].Val == nil {
+		log.Debugf("Could not find public key for %v in DHT", p)
+		return nil, routing.ErrNotFound
+	}
+
+	_, err = dht.Selector.BestRecord(pkkey, [][]byte{vals[0].Val})
```

Comment:
Is this really necessary? Doesn't `GetValues` already validate the public key? If it is, we should comment why.

Reply:
Man, yeah what was I thinking? Too much late night coding :)